transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,347 @@
+   if (close(fd) < 0) {
+       throwIOException(env, exceptionMessage("Error closing file:", errno));
+   }
+}
+
+JNIEXPORT int JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_open(JNIEnv* env, jclass clazz, jstring path) {
+
+    const char* f_path = (*env)->GetStringUTFChars(env, path, 0);
+
+    int res = open(f_path, O_RDWR | O_CREAT | O_DIRECT, 0666);
+
+    (*env)->ReleaseStringUTFChars(env, path, f_path);
+
+    if (res < 0) {
+       throwIOException(env, exceptionMessage("Cannot open file:", errno));
+    }
+
+    return res;
+}
+
+static inline void submit(JNIEnv * env, io_context_t ioContext,  struct iocb * iocb) {
+    int result = io_submit(ioContext, 1, &iocb);
+
+    if (result < 0) {
+        if (result == -EAGAIN) {
+            throwIOException(env, "Not enough space on libaio queue");
+            return;
+        }
+        else {
+            throwIOException(env, exceptionMessage("Error while submitting IO:", -result));
+            return;
+        }
+    }
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_submitWrite
+  (JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferWrite, jobject callback) {
+
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    struct iocb * iocb = getIOCB(theControl);
+
+    if (iocb == NULL) {
+       throwIOException(env, "Not enough space on the queue for submitting");
+       return;
+    }
+
+    io_prep_pwrite(iocb, fileHandle, getBuffer(env, bufferWrite), size, position);
+    iocb->data = (void *) (*env)->NewGlobalRef(env, callback);
+
+    submit(env, theControl->ioContext, iocb);
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_submitRead
+  (JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferRead, jobject callback) {
+
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    struct iocb * iocb = getIOCB(theControl);
+
+    if (iocb == NULL) {
+       throwIOException(env, "Not enough space on the queue for submitting");
+       return;
+    }
+
+    io_prep_pread(iocb, fileHandle, getBuffer(env, bufferRead), size, position);
+    iocb->data = (void *) (*env)->NewGlobalRef(env, callback);
+
+
+    submit(env, theControl->ioContext, iocb);
+}
+
+JNIEXPORT jint JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_poll
+  (JNIEnv * env, jobject obj, jobject contextPointer, jobjectArray callbacks, jint min, jint max) {
+
+    int i = 0;
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    int result = io_getevents(theControl->ioContext, min, max, theControl->events, 0);
+
+    for (i = 0; i < result; i++) {
+        struct iocb * iocbp = (struct iocb *)theControl->events[i].obj;
```

Comment:
`theControl->events[i]` is looked up multiple times.  Consider storing a pointer to this to avoid repeating array memory lookup arithmetic.